avogadro/core/atomutilities.cpp

```diff
@@ -0,0 +1,238 @@
+      for (NeighborListType::const_iterator nbIt = nbrBonds.begin(),
+                                            nbItEnd = nbrBonds.end();
+           nbIt != nbItEnd; ++nbIt) {
+        Atom a2 = getOtherAtom(a1, *nbIt);
+        if (a2.index() == atom.index())
+          continue; // we want a *new* atom
+
+        Vector3 delta = a2.position3d() - a1.position3d();
+        if (!delta.isZero(1e-5))
+          bond2 = delta.normalized();
+
+        // Check for carboxylate (CO2)
+        if ((atom.atomicNumber() == 8)  // atom for H is O
+            && (a1.atomicNumber() == 6) // central atom is C
+            && (nbIt->order() == 2) && (a2.atomicNumber() == 8))
+          break; // make sure the H will be trans to the C=O
+      }
+    }
+
+    Vector3 v1, v2;
+    v1 = bond1.cross(bond2);
+    bool noA2 = false;
+    if (bond2.norm() < 1.0e-5 || v1.norm() < 1.0e-5) {
+      //        std::cout << " creating a random paired atom " << std::endl;
+
+      // there is no a-2 atom
+      noA2 = true;
+      v2 = Vector3::Random().normalized();
+
+      double angle = fabs(acos(bond1.dot(v2)));
+      while (angle < 45.0 * DEG_TO_RAD || angle > 135.0 * DEG_TO_RAD) {
+        v2 = Vector3::Random().normalized();
+        angle = fabs(acos(bond1.dot(v2)));
+        //          std::cout << "angle = " << angle*RAD_TO_DEG << std::endl;
+      }
+      v1 = bond1.cross(v2); // so find a perpendicular, given the random vector
+      v2 = bond1.cross(v1);
+    } else {
+      //        std::cout << " found a neighbor for trans " << std::endl;
+      v1 = bond1.cross(bond2);
+      v2 = -1.0 * bond1.cross(v1);
+    }
+    v2.normalize();
+
+    switch (hybridization) {
+      case Core::SP:
+      case Core::SquarePlanar:
+      case Core::TrigonalBipyramidal:
+        newPos = bond1; // 180 degrees away from the current neighbor
+        break;
+      case Core::SP2: // sp2
+        newPos = bond1 - v2 * tan(DEG_TO_RAD * 120.0);
+        break;
+      case Core::Octahedral: // octahedral
+        newPos = bond1 - v2 * tan(DEG_TO_RAD * 90.0);
+        break;
+      case Core::SP3:
+      default:
+        newPos = (bond1 - v2 * tan(DEG_TO_RAD * M_TETRAHED));
+        break;
+    }
+
+    //      std::cout << " one bond " << newPos.normalized() << std::endl;
+    return -1.0 * newPos.normalized();
+  } // end one bond
+  else if (currentValence == 2) {
+    Vector3 bond1 = allVectors[0];
+    Vector3 bond2 = allVectors[1];
+
+    Vector3 v1 = bond1 + bond2;
+    v1.normalize();
+
+    switch (hybridization) {
+      case Core::SP: // shouldn't happen, but maybe with metal atoms?
+      case Core::SP2:
+        newPos = v1; // point away from the two existing bonds
+        break;
+      case Core::SP3:
+      default:
+        Vector3 v2 = bond1.cross(bond2); // find the perpendicular
+        v2.normalize();
+        newPos = bond1 - v2 * tan(DEG_TO_RAD * (M_TETRAHED));
+        newPos = v2 + v1 * (sqrt(2.0) / 2.0);
```

Comment:
This is definitely weird. Seems like I adapted this from Open Babel. I think we can safely comment out line 194 since it's effectively unused?

Reply:
Done!